atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/query/GetQuerySpec.java

```diff
@@ -38,67 +36,74 @@
 
     /**
      * Since each query is constructed for one cell we are using an optimisation that we can ask the CQL to do most of
-     * the work internally. First of all we are using the fact that timestamps in {@code column2} are ordered in
-     * {@code ASC} order and since we are interested in the most recent timestamp we use {@code LIMIT 1} to get the
-     * latest value. This should help with both cassandra workload and amount of transferred data.
+     * the work internally. First of all we are using the fact that cells are ordered by {@code column} and
+     * {@code timestamp} value in {@code ASC} order and since we are interested in the value with the most recent
+     * timestamp we use {@code LIMIT 1} to get the latest value. This helps with both cassandra workload and amount of
+     * transferred data.
      */
     private static final String QUERY_FORMAT = "SELECT value, column2 FROM \"%s\".\"%s\" "
             + "WHERE key = :row AND column1 = :column AND column2 > :timestamp "
-            + "ORDER BY column1, column2 ASC "
             + "LIMIT 1;";
 
-    @org.immutables.value.Value.Auxiliary
-    public abstract ByteBuffer row();
-
-    @org.immutables.value.Value.Auxiliary
-    public abstract ByteBuffer column();
+    @Override
+    public String formatQueryString() {
+        return String.format(
+                QUERY_FORMAT,
+                cqlQueryContext().keyspace(),
+                AbstractKeyValueService.internalTableName(cqlQueryContext().tableReference()));
+    }
 
-    @org.immutables.value.Value.Auxiliary
-    public abstract long humanReadableTimestamp();
+    @Override
+    public Statement makeExecutableStatement(PreparedStatement preparedStatement) {
+        return preparedStatement.bind()
+                .setBytes("row", getQueryParameters().row())
+                .setBytes("column", getQueryParameters().column())
+                .setLong("timestamp", getQueryParameters().queryTimestamp());
+    }
 
     @Override
     public ConsistencyLevel queryConsistency() {
         return ConsistencyLevel.LOCAL_QUORUM;
     }
 
     @Override
-    public final Supplier<RowStreamAccumulator<Optional<Value>>> rowStreamAccumulatorFactory() {
+    public Supplier<RowStreamAccumulator<Optional<Value>>> rowStreamAccumulatorFactory() {
```

Comment:
as an aside, why is this a `Supplier`? `GetQuerySpec` is not long-lived, you make one for every cell, so you might as well just return a `RowStreamAccumulator` directly.

Reply:
Agreed.